src/docfx/validation/JsonSchemaValidatorExtension.cs

```diff
@@ -4,31 +4,41 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.Docs.Validation;
 
 namespace Microsoft.Docs.Build
 {
-    internal class JsonSchemaValidatorExtension
+    internal class JsonSchemaValidatorExtension // todo rename
     {
         private readonly DocumentProvider _documentProvider;
-        private readonly PublishUrlMap _publishUrlMap;
+        private readonly Lazy<PublishUrlMap> _publishUrlMap;
         private readonly MonikerProvider _monikerProvider;
         private readonly ErrorBuilder _errorLog;
+        private readonly Config _config;
+        private readonly Dictionary<string, List<SourceInfo<CustomRule>>> _customRules = new Dictionary<string, List<SourceInfo<CustomRule>>>();
```

Comment:
nit: remove `SourceInfo`, make it readonly?
```suggestion
        private readonly IReadOnlyDictionary<string, CustomRule[]> _customRules;
```

Reply:
`SourceInfo` removed. For readonly, it need changes a lot even JsonSchema.Rules.